### PR TITLE
Filter form builder output by schema

### PR DESCRIPTION
**Pull request by alimpens**
# Meldingen

Ticket: [SIG-6175](https://gemeente-amsterdam.atlassian.net/browse/SIG-6175)

This PR is a bit bigger than what's outlined in the ticket, but I think it helps make this function more maintainable.
It roughly does 2 things:

- The form builder output is filtered by the schemas generated by the backend. This means that when the backend changes the schema, all we need to do is run the `generate-api-client` script to update the filter as well.
- The filter function is tested using a snapshot now. We build up a realistic mock in `filterFormResponseMock.ts` and compare that to the snapshot. If the output of the filter function changes in a way that we expect (for instance, if the backend changes the schema), all we need to do is update the snapshot. 

[SIG-6175]: https://gemeente-amsterdam.atlassian.net/browse/SIG-6175?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by vdegraaf — COMMENTED** [review]
Looks good! 
Two comments:
- Maybe a comment for firstLevelComponents and secondLevelComponents. It suggests  some fixed object schema for the input, but you cannot refer to it with `any`
- Shouldn't it be possible to provide a return type for filterFormResponse?

File: apps/admin/src/pages/form/utils/filterFormResponse.ts [hunk 1]
```diff
@@ -1,36 +1,76 @@
...
-  if (obj !== Object(obj)) {
-    return obj
-  }
+  const firstLevelComponents = obj.components.map((firstLevelComponent: any) => {
+    const secondLevelComponents = firstLevelComponent.components.map((secondLevelComponent: any) => {
+      return {
+        ...filterBySchemaPerType(secondLevelComponent),
+        validate: getFilteredValidateObject(secondLevelComponent.validate),
+      }
+    })
```
Comment (vdegraaf) [inline apps/admin/src/pages/form/utils/filterFormResponse.ts:56]: Maybe a comment for firstLevelComponents and secondLevelComponents. It suggests some fixed object schema for the input, but you cannot refer to it with any

File: apps/admin/src/pages/form/utils/filterFormResponse.ts [hunk 2]
```diff
@@ -1,36 +1,76 @@
...
+  // Explicitly remove the 'json' key if its value is an empty string, the API doesn't accept that
+  if (validate?.json === '') {
+    delete validate.json
+  }
+
+  return validate
 }
 
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
 export const filterFormResponse = (obj: any): any => {
```
Comment (vdegraaf) [inline apps/admin/src/pages/form/utils/filterFormResponse.ts:49]: Shouldn't it be possible to provide a return type for filterFormResponse?

File: apps/admin/src/pages/form/utils/filterFormResponse.ts (hunk 1 above)
Comment (alimpens) [reply apps/admin/src/pages/form/utils/filterFormResponse.ts:57]: What do you mean with 'you cannot refer to it with any'? 

In the builder, every questionnaire is a wizard with panels, which in turn contain questions. So in that sense, the schema has a fixed maximum depth of 2 (we could change that if we'd allow separate 'panel' components, but we probably won't).

With that information, could you suggest a comment? Or, preferably, a better name for first- and secondLevelComponents? I'm not sure what you're looking for.

File: apps/admin/src/pages/form/utils/filterFormResponse.ts (hunk 2 above)
Comment (alimpens) [reply apps/admin/src/pages/form/utils/filterFormResponse.ts]: Good catch, this doesn't have to be `any`. I updated it!

File: apps/admin/src/pages/form/utils/filterFormResponse.ts (hunk 1 above)
Comment (vdegraaf) [reply apps/admin/src/pages/form/utils/filterFormResponse.ts:57]: Something like:
A questionnaire can have a wizard with panels containing questions. Therefore the object has a fixed depth of two levels.

**Review by vdegraaf — APPROVED** [review]

**Pull request merged**